tests/test_2306_cppyy_git.py

```diff
@@ -0,0 +1,108 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+
+import pytest
+
+import awkward as ak
+import awkward._connect.cling
+
+cppyy = pytest.importorskip("cppyy")
+
+
+def test_array_as_generated_dataset():
+    array = ak.Array(
+        [
+            [{"x": 1, "y": [1.1]}, {"x": 2, "y": [2.2, 0.2]}],
+            [],
+            [{"x": 3, "y": [3.0, 0.3, 3.3]}],
+        ]
+    )
+
+    generator = ak._connect.cling.togenerator(array.layout.form, flatlist_as_rvec=False)
+    lookup = ak._lookup.Lookup(array.layout)
+
+    source_code = f"""
+    double go_fast(ssize_t length, ssize_t* ptrs) {{
+        auto awkward_array = {generator.dataset()};
+        double out = 0.0;
+
+        for (auto list : awkward_array) {{
+            for (auto record : list) {{
+                for (auto item : record.y()) {{
+                    out += item;
+                }}
+            }}
+        }}
+
+        return out;
+    }}
+    """
+
+    generator.generate(cppyy.cppdef)
+    cppyy.cppdef(source_code)
+    out = cppyy.gbl.go_fast(len(array), lookup.arrayptrs)
+    assert out == ak.sum(array["y"])
+
+
+def test_array_as_type():
+    array = ak.Array(
+        [
+            [{"x": 1, "y": [1.1]}, {"x": 2, "y": [2.2, 0.2]}],
+            [],
+            [{"x": 3, "y": [3.0, 0.3, 3.3]}],
+        ]
+    )
+
+    source_code_cpp = f"""
+    double go_fast_cpp({array.cpptype} awkward_array) {{
+        double out = 0.0;
+
+        for (auto list : awkward_array) {{
+            for (auto record : list) {{
+                for (auto item : record.y()) {{
+                    out += item;
+                }}
+            }}
+        }}
+
+        return out;
+    }}
+    """
+
+    cppyy.cppdef(source_code_cpp)
+
+    out = cppyy.gbl.go_fast_cpp(array.__castcpp__())
+    assert out == ak.sum(array["y"])
+
+
+def test_array_as_templated_type():
+    array = ak.Array(
+        [
+            [{"x": 1, "y": [1.1]}, {"x": 2, "y": [2.2, 0.2]}],
+            [],
+            [{"x": 3, "y": [3.0, 0.3, 3.3]}],
+        ]
+    )
+
+    # Note, 'awkward_array' of a type T has to be passed in by-value
+    source_code_cpp = """
+    template<typename T>
+    double go_fast_cpp_2(T awkward_array) {
+        double out = 0.0;
+
+        for (auto list : awkward_array) {
+            for (auto record : list) {
+                for (auto item : record.y()) {
+                    out += item;
+                }
+            }
+        }
+
+        return out;
+    }
+    """
+
+    cppyy.cppdef(source_code_cpp)
+
+    out = cppyy.gbl.go_fast_cpp_2[array.cpptype](array.__castcpp__())
+    assert out == ak.sum(array["y"])
```

Comment:
`array.__castcpp__()` returns `array.cpppars` - a Python tuple of parameters that are passed to a C++ `array.cpptype` constructor.

Reply:
I just deleted some comments I was making on the first two tests because this third example illustrates the kind of interface we want to provide (except that cppyy should automatically convert `array` into `array.__castcpp__()`).

We don't want to make users write f-strings because C++ uses curly brackets a lot, and they become `{{` and `}}`. Besides being confusing or ugly, the doubling of curly brackets makes it much more labor-intensive to copy C++ code from an f-string to a non-f-string (or C++ source file). Also, the error messages from doing the wrong thing would be surprising. It was fine for us to use in our internal code for RDataFrame, but we shouldn't require users to write that. Fortunately, C++ provides a solution: templates!

Reply:
cppyy automatically converts array into its C++ instance.